mypy/semanal.py

```diff
@@ -1469,8 +1469,8 @@ def analyze_typeddict_classdef(self, defn: ClassDef) -> bool:
             for decorator in defn.decorators:
                 decorator.accept(self)
                 if isinstance(decorator, RefExpr):
-                    if decorator.fullname in FINAL_DECORATOR_NAMES:
-                        self.fail("@final cannot be used with TypedDict", decorator)
+                    if decorator.fullname in FINAL_DECORATOR_NAMES and info is not None:
```

Comment:
What if the info is None? Does that happen if there's a circular reference? It would be good to get a test covering that code path.

Reply:
It could be `FakeInfo` for some reason.

Reply:
> What if the info is None? Does that happen if there's a circular reference?

When there's a forward declaration in the `TypedDict` body, `analyze_typeddict_classdef` gets called multiple times. `info` is initially `None` in that case, but eventually it's a `TypeInfo`.

> It would be good to get a test covering that code path.

Done.

> It could be `FakeInfo` for some reason.

When would that happen?